python/python/bystro/supervised_ppca/gf_generative_missing_pt.py

```diff
@@ -0,0 +1,326 @@
+from bystro.supervised_ppca._misc_np import (
+    softplus_inverse_np,
+    classify_missingness,
+)
+from bystro.supervised_ppca._base import BasePCASGDModel
+
+
+class PPCAM(BasePCASGDModel):
+    """
+    This implements PPCA but allowing for missing values
+    among the data. This assumes missing completely at
+    random, which is reasonable given that our lack of
+    measurements likely stems from failure to measure
+    rather than some missingness depending on the observed
+    covariates. In this case, we simply drop the missing
+    parts from the likelihood. Given that PPCA is analytic
+    in the marginal likelihood this makes fitting straightforward.
+    """
+
+    def __init__(
+        self,
+        n_components: int = 2,
+        prior_options: dict | None = None,
+        training_options: dict | None = None,
+    ):
+        """
+                Initializes the model
+
+        Parameters
+        ----------
+        n_components : int,default=2
+            The latent dimensionality
+
+        training_options : dict,default={}
+            The options for gradient descent
+
+        prior_options : dict,default={}
+            The options for priors on model parameters
+        """
+
+        super().__init__(
+            n_components=n_components,
+            prior_options=prior_options,
+            training_options=training_options,
+        )
+        self._initialize_save_losses()
+
+        self.W_: NDArray[np.float_] | None = None
+        self.sigmas_: NDArray[np.float_] | None = None
+        self.sigma2_: np.float_ | None = None
+        self.p: int | None = None
+
+    def __repr__(self):
+        return f"PPCAM(n_components={self.n_components})"
+
+    def fit(self, X: NDArray[np.float_], progress_bar: bool = True,) -> "PPCAM":
+        """
+        Fits a model given covariates X as well as option labels y in the
+        supervised methods
+
+        Parameters
+        ----------
+        X : NDArray,(n_samples,n_covariates)
+            The data
+
+        progress_bar : bool,default=True
+            Whether to print the progress bar to monitor time
+
+        Returns
+        -------
+        self : PPCA
+            The model
+        """
+        self._test_inputs(X)
+        training_options = self.training_options
+        N, p = X.shape
+        self.p = p
+
+        W_, sigmal_ = self._initialize_variables(X)
+
+        X_list, miss_pat = classify_missingness(X)
+        Xt_list = [torch.tensor(X) for X in X_list]
+
+        n_groups = len(Xt_list)
+        p_list = np.array([np.sum(mp) for mp in miss_pat])
+
+        trainable_variables = [W_, sigmal_]
+
+        optimizer = torch.optim.SGD(
+            trainable_variables,
+            lr=training_options["learning_rate"],
+            momentum=training_options["momentum"],
+        )
+        eye = torch.tensor(np.eye(p).astype(np.float32))
+        softplus = nn.Softplus()
+
+        _prior = self._create_prior()
+
+        for i in trange(
```

Comment:
Let's optimize this loop a bit

```python
for i in trange(
    training_options["n_iterations"], disable=not progress_bar
):
    sigma = softplus(sigmal_)
    WWT = torch.matmul(torch.transpose(W_, 0, 1), W_)
    Sigma = WWT + sigma * eye

   like_marginal = []
   for y in range(n_groups):
      sigma = Sigma[miss_pat[y]][:, miss_pat[y]]
      mvn = MultivariateNormal(torch.zeros(p_list[y]), sigma)
      like_marginal.append( torch.sum(mvn.log_prob(Xt_list[y])) )
   
    like_tot = torch.sum(torch.stack(like_marg)) / N
    like_prior = _prior(trainable_variables)
    posterior = like_tot + like_prior / N
    loss = -1 * posterior

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    self._save_losses(i, like_tot, like_prior, posterior)
```

Reply:
Instead of having 3 loops over n_groups, combine them into 1

Reply:
There is probably a vectorized approach as well:

‘’’py

# Assuming Sigma, Xt_list, p_list, and N are already defined

# Step 1: Create batched mean and covariance matrices
batch_means = torch.stack([torch.zeros(p_list[y]) for y in range(n_groups)])
batch_covariances = torch.stack([Sigma[miss_pat[y]][:, miss_pat[y]] for y in range(n_groups)])

# Step 2: Create a batch of multivariate normal distributions
batch_mvn = MultivariateNormal(batch_means, batch_covariances)

# Step 3: Calculate log probabilities for each group's data in a vectorized way
log_probs = batch_mvn.log_prob(torch.stack(Xt_list))

# Step 4: Sum up these log probabilities and divide by N
like_tot = torch.sum(log_probs) / N

‘’’